### ISSUE TITLE
Error installing ffi

### ISSUE THREAD

**Issue by deivid-rodriguez** [oneclick/rubyinstaller2#68]
## What problems are you experiencing?
My CI environment is failing to install `ffi`. I'm getting this error:

```
OpenSSL::SSL::SSLError: SSL_read: cert already in hash table
An error occurred while installing ffi (1.9.18), and Bundler cannot continue.
Make sure that `gem install ffi -v '1.9.18'` succeeds before bundling.
In Gemfile:
  overcommit was resolved to 0.39.1, which depends on
    childprocess was resolved to 0.6.3, which depends on
      ffi
Command exited with code 5
```

This is the relevant build log https://ci.appveyor.com/project/deivid-rodriguez/byebug/build/740/job/mvd949l4jwqdsrd1#L92

## Steps to reproduce

Unfortunately I haven't been able to reproduce locally. I wonder if it's because I'm running RubyInstaller-2.4.1-2, and appveyor might be running the older version?

**Comment by MSP-Greg** [comment]
@deivid-rodriguez 

> I wonder if it's because I'm running RubyInstaller-2.4.1-2, and appveyor might be running the older version?

Yes, it is, but I don't know if it's the cause.  I haven't had a chance to look your CI.  I will, unless Lars finds the issue.  I did a gem uninstall/gem install ffi on my trunk build, and it worked fine.

I've got a daily run on [Appveyor](https://ci.appveyor.com/project/MSP-Greg/appveyor-ruby) that shows quite a bit of info re it's ruby versions and MSYS2/MinGW packages.  Appveyor is using 2.4.1-1, the cert file is dated 25-May.

**Comment by MSP-Greg** [comment]
@deivid-rodriguez 

I believe the solution may be something similar to
```sh
    # see https://github.com/bundler/bundler/issues/5344
  - bundle config force_ruby_platform true
    # not sure if next two lines are required, I have the option set in my .gemrc file
  - bundle config build.ffi   --use-system-libraries
  - bundle config rainbow.ffi --use-system-libraries
  - bundle install --jobs=3 --retry=3
```

I ran it with three MSYS2/MinGW builds, results [here](https://ci.appveyor.com/project/MSP-Greg/byebug/build/1), using [appveyor.yml](https://github.com/MSP-Greg/byebug/blob/9.1.0_release/appveyor.yml).  The 2.4.1 build is standard Appveyor (RubyInstaller2), the 2.3.4 and trunk builds are my own.

Two items:
1. Two failures in trunk see [here](https://ci.appveyor.com/project/MSP-Greg/byebug/build/job/4uql8s08ecjka5lu).
2. For quite a while, I've done my builds with a readline extension, somewhat just to see if it would test ok.  Both of them run very slowly on the tests.  Could a 'bad' readline extension be the cause?  Maybe it's time for me to change back to rb-readline.  If not, any ideas?  I'm going to do some local testing...

My builds:

2.3.4u - installed bundler, updated RubyGems, and installed four default gems, bigdecimal, json, openssl, and psych, which are 'pushed back' into ruby.

trunk - I have been keeping the trunk build current, normally no more than (approx) 10 commits behind of trunk...

**Comment by larskanis** [comment]
Thank you for your bug report! This appears to be a bug in ruby-openssl. Puppet used the following work around: https://github.com/puppetlabs/puppet/pull/6011

Ruby-2.4.2 will ship an updated openssl gem, when it's out. And I'll notify the guys at appveyor when the RubyInstaller can be updated.

**Comment by MSP-Greg** [comment]
I believe code has been added to [byebug](https://github.com/deivid-rodriguez/byebug) such that [Appveyor](https://ci.appveyor.com/project/deivid-rodriguez/byebug) is  passing for 32 & 64 bit RI2 2.4.1.

@deivid-rodriguez are you ok to close? If so...

**Comment by deivid-rodriguez** [comment]
Yes, closing. This is not a problem in rubyinstaller2 in any case!

Thanks to both for the quick and helpful tips!

**Comment by Iristyle** [comment]
Just an FYI on this issue.

The OpenSSL 2.0.4 patch landed on 8/8/2017 in the Ruby 2.4 branch at https://github.com/ruby/ruby/commit/b08e0ade5baacbc7d4754e9f6afe8d09a3e00bc5#diff-51a5e2412fcf851e37d1883b1ae53de5 (based on ticket https://bugs.ruby-lang.org/issues/11033)

As @larskanis mentioned, we're still waiting on a new `2.4.2` tag from the Ruby folks (and a rollout to AppVeyor) so that no workaround is necessary.

Recently a new error started showing up that is intermittently causing AppVeyor failures (AppVeyor failure detection can be a bit flaky at times).  Occasionally we'll see this message in red from during AppVeyor setup (despite it just being a warning) like:

```
gem : warning: mingw-w64-x86_64-openssl-1.0.2.k-1 is up to date -- skipping
At line:17 char:5
+     gem install openssl --version '~> 2.0.4' --no-ri --no-rdoc -- --w ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (warning: mingw-...ate -- skipping:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```

I believe this message is coming from the underlying msys2 package manager (pacman), but I'm not yet sure how to disable it.  I have a PR up against Puppet where I'm working on it - https://github.com/puppetlabs/puppet/pull/6192

**Comment by MSP-Greg** [comment]
@Iristyle

> that is intermittently causing AppVeyor failures (AppVeyor failure detection can be a bit flaky at times)

I don't recall if there's a distinction between AppVeyor failures in `before_test` or `test_script`...

A standard MSYS2/MinGW install doesn't have an OpenSSL package, but Appveyor does (see [appveyor-ruby](https://ci.appveyor.com/project/MSP-Greg/appveyor-ruby)).  The OpenSSL gemspec has the the line
```
spec.metadata["msys2_mingw_dependencies"] = "openssl"
```
which is used by the RubyInstaller2 runtime, causing it install the OpenSSL package.  The 'error message'  is confirming it's action or lack thereof.

There has been some discussion about how this should work (user override?), but causing issues like this wasn't forseen...

**Comment by Iristyle** [comment]
@MSP-Greg As for the AppVeyor flakiness, In both cases, the `gem install` occurs during `test_script`. AppVeyor sometimes ignores the `gem` output warning, like in this case:
https://ci.appveyor.com/project/puppetlabs/puppet/build/4.1.0.6243/job/wtm0yegtgxbbas1n

But other times it treats it as a failure, like here:
https://ci.appveyor.com/project/puppetlabs/puppet/build/4.1.0.6227/job/guc3gsd88d36w49h

Note line 462 at the end of the run in particular:

```
Command executed with exception: warning: mingw-w64-x86_64-openssl-1.0.2.k-1 is up to date -- skipping
```

Unfortunately this causes confusion amongst my colleagues who are not familiar with the issue (the red is usually taken as a clear indication of where to point the finger ;0).

Thanks for the pointer on the `msys2_mingw_dependencies`.  I had been looking at ways to suppress pacman warnings, gem warnings, etc.  I may be able to just reset `$LASTEXITCODE`, even though it's a bit of a hack.  I'll look into seeing if there's way I can otherwise influence the dependency resolution to curb the issue.

**Comment by Iristyle** [comment]
So the most effective "fix" (for Puppets situation, given we're running in PowerShell) looks to be to redirect stderr into stdout so that PowerShell doesn't autogenerate ErrorRecord instances for the stderr output.  The presence of ErrorRecord instances is what AppVeyor is considering a failure.  The PR I referenced at https://github.com/puppetlabs/puppet/pull/6192 contains the workaround for this situation (if we were running inside cmd.exe this wouldn't be a problem).

**Comment by larskanis** [comment]
I released [RubyInstaller-2.4.2-2](http://localhost:4000/2017/09/15/rubyinstaller-2.4.2-2-released.html) today and filed [an issue on appveyor](https://github.com/appveyor/ci/issues/1790) to update the appveyor pre-installed version.
